proxyd/latest_block_poller.go

```diff
@@ -0,0 +1,78 @@
+package proxyd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type LatestBlockPoller struct {
+	rt       RoundTripper
+	bn       atomic.Uint64
+	shutdown bool
+	mutex    sync.Mutex
+}
+
+type RoundTripper func(ctx context.Context, req json.RawMessage) (*RPCRes, error)
+
+// NewLatestBlockPoller creates a new LatestBlockPoller and starts polling
+// for the latest block number in a separate goroutine.
+func NewLatestBlockPoller(rt RoundTripper) *LatestBlockPoller {
+	p := &LatestBlockPoller{
+		rt: rt,
+	}
+	p.poll()
+	go p.start()
+	return p
+}
+
+// Get returns the latest block number.
+func (p *LatestBlockPoller) Get() uint64 {
+	return p.bn.Load()
+}
+
+// Shutdown stops the poller.
+func (p *LatestBlockPoller) Shutdown() {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	p.shutdown = true
+}
+
+func (p *LatestBlockPoller) start() {
+	ticker := time.NewTicker(2 * time.Second)
+	for range ticker.C {
+		p.mutex.Lock()
+		if p.shutdown {
+			ticker.Stop()
+			p.mutex.Unlock()
+			return
+		}
+		p.poll()
+		p.mutex.Unlock()
+	}
```

Comment:
Note, a `for range` loop over a receiving channel will _not_ execute until the first interval is passed. In other words, it will first wait 2 seconds and then perform the first poll.

To instead have this execute immediately, we could do something like the following, which will run the loop immediately and wait on receive for each subsequent iteration.
```suggestion
    for ; true; <-ticker.C {
		p.mutex.Lock()
		if p.shutdown {
			ticker.Stop()
			p.mutex.Unlock()
			return
		}
		p.poll()
		p.mutex.Unlock()
	}
```

Reply:
It's a minor thing, but i made the `poll` call in the original goroutine (and allowed the interval to pass in the separate one) so that the latest block has a chance to be populated before returning. But given the handling around 0 blocks, it probably doesn't matter too much. Let me know if you think we should switch to your suggestion.

Reply:
I lean to this being fine. The majority of the time proxyd is running, it will have this value populated, & if it doesn't it's not the end of the world.

Reply:
sgtm, will update